File: frontend/src/api/suppliers.ts
import { request, requestCollection } from "./client";
import { ContactHistoryEntry } from "./types";

export type SupplierContact = {
  name?: string | null;
  position?: string | null;
  email?: string | null;
  phone?: string | null;
  notes?: string | null;
};

export type Supplier = {
  id: number;
  name: string;
  rtn?: string | null;
  payment_terms?: string | null;
  contact_name?: string | null;
  email?: string | null;
  phone?: string | null;
  address?: string | null;
  notes?: string | null;
  outstanding_debt: number;
  history: ContactHistoryEntry[];
  contact_info: SupplierContact[];
  products_supplied: string[];
  created_at: string;
  updated_at: string;
};

export type SupplierPayload = {
  name: string;
  rtn?: string | null;
  payment_terms?: string | null;
  contact_name?: string | null;
  email?: string | null;
  phone?: string | null;
  address?: string | null;
  notes?: string | null;
  outstanding_debt?: number;
  history?: ContactHistoryEntry[];
  contact_info?: SupplierContact[];
  products_supplied?: string[];
};

export type SupplierBatch = {
  id: number;
  supplier_id: number;
  store_id?: number | null;
  device_id?: number | null;
  model_name: string;
  batch_code: string;
  unit_cost: number;
  quantity: number;
  purchase_date: string;
  notes?: string | null;
  created_at: string;
  updated_at: string;
};

export type SupplierBatchPayload = {
  store_id?: number | null;
  device_id?: number | null;
  model_name: string;
  batch_code: string;
  unit_cost: number;
  quantity?: number;
  purchase_date: string;
  notes?: string | null;
};

export type SupplierBatchOverviewItem = {
  supplier_id: number;
  supplier_name: string;
  batch_count: number;
  total_quantity: number;
  total_value: number;
  latest_purchase_date: string;
  latest_batch_code?: string | null;
  latest_unit_cost?: number | null;
};

export type SupplierAccountsPayableBucket = {
  label: string;
  days_from: number;
  days_to: number | null;
  amount: number;
  percentage: number;
  count: number;
};

export type SupplierAccountsPayableSupplier = {
  supplier_id: number;
  supplier_name: string;
  rtn?: string | null;
  payment_terms?: string | null;
  outstanding_debt: number;
  bucket_label: string;
  bucket_from: number;
  bucket_to: number | null;
  days_outstanding: number;
  last_activity?: string | null;
  contact_name?: string | null;
  contact_email?: string | null;
  contact_phone?: string | null;
  products_supplied: string[];
  contact_info: SupplierContact[];
};

export type SupplierAccountsPayableSummary = {
  total_balance: number;
  total_overdue: number;
  supplier_count: number;
  generated_at: string;
  buckets: SupplierAccountsPayableBucket[];
};

export type SupplierAccountsPayableResponse = {
  summary: SupplierAccountsPayableSummary;
  suppliers: SupplierAccountsPayableSupplier[];
};

export function listSuppliers(
  token: string,
  query?: string,
  limit = 100
): Promise<Supplier[]> {
  const params = new URLSearchParams({ limit: String(limit) });
  if (query) {
    params.append("q", query);
  }
  return requestCollection<Supplier>(`/suppliers?${params.toString()}`, { method: "GET" }, token);
}

export function getSuppliersAccountsPayable(
  token: string
): Promise<SupplierAccountsPayableResponse> {
  return request<SupplierAccountsPayableResponse>(
    "/suppliers/accounts-payable",
    { method: "GET" },
    token
  );
}

export function exportSuppliersCsv(token: string, query?: string): Promise<Blob> {
  const params = new URLSearchParams({ export: "csv" });
  if (query) {
    params.append("q", query);
  }
  return request<Blob>(`/suppliers?${params.toString()}`, { method: "GET" }, token);
}

export function createSupplier(
  token: string,
  payload: SupplierPayload,
  reason: string
): Promise<Supplier> {
  return request<Supplier>(
    "/suppliers",
    { method: "POST", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token
  );
}

export function updateSupplier(
  token: string,
  supplierId: number,
  payload: Partial<SupplierPayload>,
  reason: string
): Promise<Supplier> {
  return request<Supplier>(
    `/suppliers/${supplierId}`,
    { method: "PUT", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token
  );
}

export function deleteSupplier(token: string, supplierId: number, reason: string): Promise<void> {
  return request<void>(
    `/suppliers/${supplierId}`,
    { method: "DELETE", headers: { "X-Reason": reason } },
    token
  );
}

export function getSupplierBatchOverview(
  token: string,
  storeId: number,
  limit = 5,
): Promise<SupplierBatchOverviewItem[]> {
  return requestCollection<SupplierBatchOverviewItem>(
    `/reports/inventory/supplier-batches?store_id=${storeId}&limit=${limit}`,
    { method: "GET" },
    token,
  );
}

export function listSupplierBatches(
  token: string,
  supplierId: number,
  limit = 50
): Promise<SupplierBatch[]> {
  const params = new URLSearchParams({ limit: String(limit) });
  return requestCollection<SupplierBatch>(
    `/suppliers/${supplierId}/batches?${params.toString()}`,
    { method: "GET" },
    token
  );
}

export function createSupplierBatch(
  token: string,
  supplierId: number,
  payload: SupplierBatchPayload,
  reason: string
): Promise<SupplierBatch> {
  return request<SupplierBatch>(
    `/suppliers/${supplierId}/batches`,
    { method: "POST", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token
  );
}

export function updateSupplierBatch(
  token: string,
  batchId: number,
  payload: Partial<SupplierBatchPayload>,
  reason: string
): Promise<SupplierBatch> {
  return request<SupplierBatch>(
    `/suppliers/batches/${batchId}`,
    { method: "PUT", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token
  );
}

export function deleteSupplierBatch(token: string, batchId: number, reason: string): Promise<void> {
  return request<void>(
    `/suppliers/batches/${batchId}`,
    { method: "DELETE", headers: { "X-Reason": reason } },
    token
  );
}

File: backend/app/schemas/suppliers.py
"""Esquemas para la gestión de proveedores."""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Iterable

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator

from .common import ContactHistoryEntry, normalize_optional_rtn_value


class SupplierContact(BaseModel):
    name: str | None = Field(default=None, max_length=120)
    position: str | None = Field(default=None, max_length=120)
    email: str | None = Field(default=None, max_length=120)
    phone: str | None = Field(default=None, max_length=40)
    notes: str | None = Field(default=None, max_length=255)

    @field_validator("name", "position", "email", "phone", "notes", mode="before")
    @classmethod
    def _normalize_contact_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class SupplierBase(BaseModel):
    rtn: str | None = Field(default=None, max_length=30)
    payment_terms: str | None = Field(default=None, max_length=80)
    contact_name: str | None = Field(default=None, max_length=120)
    email: str | None = Field(default=None, max_length=120)
    phone: str | None = Field(default=None, max_length=40)
    address: str | None = Field(default=None, max_length=255)
    notes: str | None = Field(default=None, max_length=500)
    outstanding_debt: Decimal = Field(default=Decimal("0"))
    history: list[ContactHistoryEntry] = Field(default_factory=list)
    contact_info: list[SupplierContact] = Field(default_factory=list)
    products_supplied: list[str] = Field(default_factory=list)

    @field_validator(
        "rtn",
        "payment_terms",
        "contact_name",
        "email",
        "phone",
        "address",
        "notes",
        mode="before",
    )
    @classmethod
    def _normalize_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @field_validator("rtn", mode="before")
    @classmethod
    def _normalize_rtn(cls, value: str | None) -> str | None:
        return normalize_optional_rtn_value(value)

    @field_validator("products_supplied", mode="before")
    @classmethod
    def _normalize_products(cls, value: object) -> list[str]:
        if value is None:
            return []
        if isinstance(value, (str, bytes)):
            candidates = [value]
        else:
            candidates = list(value) if isinstance(
                value, Iterable) else [value]
        normalized: list[str] = []
        for candidate in candidates:
            if candidate is None:
                continue
            text = str(candidate).strip()
            if not text:
                continue
            if text not in normalized:
                normalized.append(text)
        return normalized

    @field_serializer("outstanding_debt")
    @classmethod
    def _serialize_debt(cls, value: Decimal) -> float:
        return float(value)


class SupplierCreate(SupplierBase):
    name: str = Field(..., max_length=120)

    @field_validator("name")
    @classmethod
    def _normalize_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("El nombre es obligatorio.")
        return normalized


class SupplierUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=120)
    rtn: str | None = Field(default=None, max_length=30)
    payment_terms: str | None = Field(default=None, max_length=80)
    contact_name: str | None = Field(default=None, max_length=120)
    email: str | None = Field(default=None, max_length=120)
    phone: str | None = Field(default=None, max_length=40)
    address: str | None = Field(default=None, max_length=255)
    notes: str | None = Field(default=None, max_length=500)
    outstanding_debt: Decimal | None = Field(default=None)
    history: list[ContactHistoryEntry] | None = Field(default=None)
    contact_info: list[SupplierContact] | None = Field(default=None)
    products_supplied: list[str] | None = Field(default=None)

    @field_validator(
        "name",
        "rtn",
        "payment_terms",
        "contact_name",
        "email",
        "phone",
        "address",
        "notes",
        mode="before",
    )
    @classmethod
    def _normalize_update_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @field_validator("rtn", mode="before")
    @classmethod
    def _normalize_update_rtn(cls, value: str | None) -> str | None:
        return normalize_optional_rtn_value(value)

    @field_validator("products_supplied", mode="before")
    @classmethod
    def _normalize_products_update(cls, value: object) -> list[str] | None:
        if value is None:
            return None
        if isinstance(value, (str, bytes)):
            candidates = [value]
        else:
            candidates = list(value) if isinstance(
                value, Iterable) else [value]
        normalized: list[str] = []
        for candidate in candidates:
            if candidate is None:
                continue
            text = str(candidate).strip()
            if not text:
                continue
            if text not in normalized:
                normalized.append(text)
        return normalized


class SupplierResponse(SupplierBase):
    id: int
    name: str
    created_at: datetime
    updated_at: datetime
    is_deleted: bool = Field(default=False)
    deleted_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class SupplierAccountsPayableBucket(BaseModel):
    label: str
    days_from: int
    days_to: int | None
    amount: float
    percentage: float
    count: int


class SupplierAccountsPayableSupplier(BaseModel):
    supplier_id: int
    supplier_name: str
    rtn: str | None
    payment_terms: str | None
    outstanding_debt: float
    bucket_label: str
    bucket_from: int
    bucket_to: int | None
    days_outstanding: int
    last_activity: datetime | None
    contact_name: str | None
    contact_email: str | None
    contact_phone: str | None
    products_supplied: list[str]
    contact_info: list[SupplierContact] = Field(default_factory=list)


class SupplierAccountsPayableSummary(BaseModel):
    total_balance: float
    total_overdue: float
    supplier_count: int
    generated_at: datetime
    buckets: list[SupplierAccountsPayableBucket]


class SupplierAccountsPayableResponse(BaseModel):
    summary: SupplierAccountsPayableSummary
    suppliers: list[SupplierAccountsPayableSupplier]


class SupplierBatchBase(BaseModel):
    model_name: str = Field(..., max_length=120)
    batch_code: str = Field(..., max_length=80)
    unit_cost: Decimal = Field(..., ge=Decimal("0"))
    quantity: int = Field(default=0, ge=0)
    purchase_date: date
    notes: str | None = Field(default=None, max_length=255)
    store_id: int | None = Field(default=None, ge=1)
    device_id: int | None = Field(default=None, ge=1)

    model_config = ConfigDict(protected_namespaces=())

    @field_validator("model_name", "batch_code", "notes", mode="before")
    @classmethod
    def _normalize_batch_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @field_serializer("unit_cost")
    @classmethod
    def _serialize_unit_cost(cls, value: Decimal) -> float:
        return float(value)


class SupplierBatchCreate(SupplierBatchBase):
    """Datos requeridos para registrar un nuevo lote de proveedor."""


class SupplierBatchUpdate(BaseModel):
    model_name: str | None = Field(default=None, max_length=120)
    batch_code: str | None = Field(default=None, max_length=80)
    unit_cost: Decimal | None = Field(default=None, ge=Decimal("0"))
    quantity: int | None = Field(default=None, ge=0)
    purchase_date: date | None = None
    notes: str | None = Field(default=None, max_length=255)
    store_id: int | None = Field(default=None, ge=1)
    device_id: int | None = Field(default=None, ge=1)

    model_config = ConfigDict(protected_namespaces=())

    @field_validator("model_name", "batch_code", "notes", mode="before")
    @classmethod
    def _normalize_optional_batch_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class SupplierBatchResponse(SupplierBatchBase):
    id: int
    supplier_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class SupplierBatchOverviewItem(BaseModel):
    supplier_id: int
    supplier_name: str
    batch_count: int = Field(ge=0)
    total_quantity: int = Field(ge=0)
    total_value: float = Field(ge=0)
    latest_purchase_date: date
    latest_batch_code: str | None = None
    latest_unit_cost: float | None = Field(default=None, ge=0)
